svf-llvm/CMakeLists.txt

```diff
@@ -1,61 +1,129 @@
+  Build type    ${LLVM_BUILD_TYPE}
+  Exceptions    ${LLVM_ENABLE_EH}
+  RTTI          ${LLVM_ENABLE_RTTI}
+  Dynamic lib   ${LLVM_LINK_LLVM_DYLIB}"
+)
+
+if(CMAKE_BUILD_TYPE STREQUAL "Debug" AND NOT ${LLVM_BUILD_TYPE} STREQUAL "Debug")
+  message(NOTICE "Building SVF in debug-mode but LLVM was not built in debug-mode; "
+                  "debug information could be incomplete when using SVF from LLVM")
+endif()
+
+include_directories(SYSTEM ${LLVM_INCLUDE_DIRS})
+link_directories(${LLVM_LIBRARY_DIRS})
+add_definitions(${LLVM_DEFINITIONS})
+
+if(NOT LLVM_ENABLE_EH)
+  message(STATUS "Building SVF without exception handling")
+  set(SVF_NO_RTTI ON PARENT_SCOPE)  # Set for parent scope because this needs to be set in top-level CMakeLists.txt
+endif()
 
-  list(APPEND CMAKE_MODULE_PATH "${LLVM_CMAKE_DIR}")
-  include(AddLLVM)
+if(NOT LLVM_ENABLE_RTTI)
+  message(STATUS "Building SVF without RTTI")
+  set(SVF_NO_EH ON PARENT_SCOPE)  # Set for parent scope because this needs to be set in top-level CMakeLists.txt
+endif()
 
-  add_definitions(${LLVM_DEFINITIONS})
-  # include_directories(SYSTEM ${LLVM_INCLUDE_DIRS})
+if(LLVM_LINK_LLVM_DYLIB)
+  message(STATUS "Linking to LLVM dynamic shared library object")
+  set(llvm_libs LLVM)
 
-  if(LLVM_LINK_LLVM_DYLIB)
-    set(llvm_libs LLVM)
+  # Set which components to include in the dynamic library to include the new SvfLLVM
+  if (LLVM_DYLIB_COMPONENTS)
+    message(STATUS "Appending SvfLLVM to LLVM dynamic library components")
+    list(APPEND LLVM_DYLIB_COMPONENTS SvfLLVM)
   else()
-    llvm_map_components_to_libnames(
-      llvm_libs
-      analysis
-      bitwriter
-      core
-      instcombine
-      instrumentation
-      ipo
-      irreader
-      linker
-      scalaropts
-      support
-      target
-      transformutils)
+    message(STATUS "Adding all;SvfLLVM to LLVM dynamic library components (was unset)")
+    set(LLVM_DYLIB_COMPONENTS all;SvfLLVM)
   endif()
 else()
-  set(IN_SOURCE_BUILD 1)
+  message(STATUS "Linking to separate LLVM static libraries")
+  llvm_map_components_to_libnames(
+    llvm_libs
+    analysis
+    bitwriter
+    core
+    instcombine
+    instrumentation
+    ipo
+    irreader
+    linker
+    scalaropts
+    support
+    target
+    transformutils
+)
 endif()
 
-# SVF-LLVM contains LLVM Libs
-file(GLOB SVFLLVM_SOURCES lib/*.cpp)
+# Make the "add_llvm_library()" command available and configure LLVM/CMake
+list(APPEND CMAKE_MODULE_PATH "${LLVM_CMAKE_DIR}")
+include(AddLLVM)
 
 add_llvm_library(SvfLLVM ${SVFLLVM_SOURCES})
-target_include_directories(SvfLLVM SYSTEM PUBLIC ${LLVM_INCLUDE_DIRS})
 target_include_directories(SvfLLVM PUBLIC include)
 target_link_libraries(SvfLLVM PUBLIC ${llvm_libs} ${Z3_LIBRARIES} SvfCore)
 
-if(DEFINED IN_SOURCE_BUILD)
+# Build SVF with C++ standard C++17
+set_property(TARGET SvfLLVM PROPERTY CXX_STANDARD 17)
+set_property(TARGET SvfLLVM PROPERTY CXX_EXTENSIONS ON)
+set_property(TARGET SvfLLVM PROPERTY CXX_STANDARD_REQUIRED ON)
```

Comment:
Any better reason to set the C++ standard for SVFcore and SVF-LLVM separately? I think the original way is less repetitive.

Reply:
I think CMake generally recommends applying options/properties to specific targets rather than using global options (e.g. the documentation for [`link_libraries()`](https://cmake.org/cmake/help/latest/command/link_libraries.html) states: "*The [target_link_libraries()](https://cmake.org/cmake/help/latest/command/target_link_libraries.html#command:target_link_libraries) command should be preferred whenever possible*".

However, I think the `set_property()` command -- by default -- applies the properties to the current directory and any subdirectories included later. I applied the option to each target separately to avoid clobbering the build configuration of any projects that might include SVF through `add_subdirectory()`, but I think this is indeed unnecessary. I'll move this back to the top-level `CMakeLists.txt`.